ai/src/algorithms/sacd/agent.py

```diff
@@ -0,0 +1,188 @@
+        for critic_var, target_var in zip(self._critic_2_variables, self._target_critic_2_variables):
+            target_var.assign(self.TARGET_MOMEMTUM * critic_var + (1.0 - self.TARGET_MOMEMTUM) * target_var)
+
+    def train(self, data):
+        state_batch, action_batch, reward_batch, next_state_batch = data
+        action_batch = tf.cast(tf.expand_dims(action_batch, 1), tf.float64)
+        reward_batch = tf.cast(tf.expand_dims(reward_batch, 1), tf.float64)
+
+        with tf.name_scope('actor_loss'):
+            with tf.GradientTape() as actor_tape:
+                _action, action_probs = self.actor(state_batch)
+                action_logprobs = tf.math.log(action_probs)
+                q1_value = self._critic_1(state_batch)
+                q2_value = self._critic_2(state_batch)
+                q_log_target = tf.minimum(q1_value, q2_value)
+                actor_loss = tf.reduce_mean(
+                    tf.reduce_sum(action_probs * (self.alpha * action_logprobs - q_log_target), 1))
+        self._actor_optimizer.apply_gradients(
+            zip(actor_tape.gradient(actor_loss, self._actor_variables), self._actor_variables))
+
+        with tf.name_scope('critic_loss'):
+            _next_action, next_action_probs = self.actor(next_state_batch)
+            next_action_logprobs = tf.math.log(next_action_probs)
+            q1_next_target = self._critic_1(next_state_batch)
+            q2_next_target = self._critic_2(next_state_batch)
+            min_q = next_action_probs * (
+                tf.minimum(q1_next_target, q2_next_target) - self.alpha * next_action_logprobs)
+            q_target = reward_batch + self.REWARD_DISCOUNT * min_q
+
+            critic_losses = []
+            critic_tapes = []
+            for q_net in [self._critic_1, self._critic_2]:
+                with tf.GradientTape() as critic_tape:
+                    q_value = tf.gather(q_net(state_batch), tf.cast(action_batch, tf.int64), axis=1)
+                    critic_losses.append(0.5 * tf.reduce_mean((q_value - q_target) ** 2))
+                critic_tapes.append(critic_tape)
+        self._critic_1_optimizer.apply_gradients(
+            zip(critic_tapes[0].gradient(critic_losses[0], self._critic_1_variables), self._critic_1_variables))
+        self._critic_2_optimizer.apply_gradients(
+            zip(critic_tapes[1].gradient(critic_losses[1], self._critic_2_variables), self._critic_2_variables))
+
+        with tf.name_scope('alpha_loss'):
+            neg_entropy = tf.reduce_sum(action_logprobs * action_probs, axis=1)
+            with tf.GradientTape() as alpha_tape:
+                alpha_loss = tf.reduce_mean(-1 * self.log_alpha * (neg_entropy + self.target_entropy))
+        self._alpha_optimizer.apply_gradients(
+            zip(alpha_tape.gradient(alpha_loss, [self.log_alpha]), [self.log_alpha]))
+        self.alpha = tf.exp(self.log_alpha)
+
+        self._copy_target_models()
+
+
+class SACDAgent(SpiceAIAgent):
```

Comment:
```suggestion
class SoftActorCriticDiscreteAgent(SpiceAIAgent):
```

I would suggest making this explicit - someone reading the code in the factory method won't know what SACD stands for.

Reply:
Yes good idea